File: Proyecto/Calendario_App/src/main/java/org/example/calendario_app/dao/impl/ClienteDAOImpl.java
package org.example.calendario_app.dao.impl;

import org.example.calendario_app.model.Cliente;
import org.example.calendario_app.util.DatabaseConnection;

import java.sql.*;

public class ClienteDAOImpl {
    DatabaseConnection databaseConnection;

    public ClienteDAOImpl() {
        databaseConnection = new DatabaseConnection();
    }

    public int registrar(Cliente cliente) {
        String query = "INSERT INTO clientes (nombre, apellidos) VALUES (?, ?)";
        int idGenerado = -1;

        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {

            pstmt.setString(1, cliente.getNombre());
            pstmt.setString(2, cliente.getApellidos());

            int affectedRows = pstmt.executeUpdate();

            if (affectedRows > 0) {
                try (ResultSet generatedKeys = pstmt.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        idGenerado = generatedKeys.getInt(1);
                    }
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return idGenerado;
    }

    public Cliente obtenerPorId(int id) {
        String query = "SELECT * FROM clientes WHERE id_cliente = ?";
        Cliente cliente = null;
        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query)) {
            pstmt.setInt(1, id);
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    cliente = new Cliente(
                            rs.getString("nombre"),
                            rs.getString("apellidos"),
                            null);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return cliente;
    }
}

File: Proyecto_Android/Calendario_Android_App/app/src/main/java/com/example/calendario_android_app/dao/impl/ClienteDAOImpl.kt
package com.example.calendario_android_app.dao.impl

import com.example.calendario_android_app.dao.ClienteDAO
import com.example.calendario_android_app.model.Cliente
import com.example.calendario_android_app.util.DatabaseConnection
import java.sql.SQLException

class ClienteDAOImpl : ClienteDAO {
    override fun obtenerCliente(id_cliente: Int): Cliente? {
        val connection = DatabaseConnection.getConnection() ?: return null
        var cliente: Cliente? = null

        try {
            val query = "SELECT * FROM clientes WHERE id_cliente = ?"
            val statement = connection.prepareStatement(query)
            statement.setInt(1, id_cliente)
            
            val resultSet = statement.executeQuery()
            
            if (resultSet.next()) {
                cliente = Cliente(
                    id_cliente = resultSet.getInt("id_cliente"),
                    nombre = resultSet.getString("nombre"),
                    apellidos = resultSet.getString("apellidos"),
                    creado_el = resultSet.getTimestamp("creado_el"),
                    actualizado_el = resultSet.getTimestamp("actualizado_el")
                )
            }
            resultSet.close()
            statement.close()
            connection.close()
        } catch (e: SQLException) {
            android.util.Log.e("ClienteDAOImpl", "SQL Error", e)
            e.printStackTrace()
        }
        return cliente
    }

    override fun crearCliente(nombre: String, apellidos: String): Int? {
        val connection = DatabaseConnection.getConnection() ?: return null
        var newId: Int? = null

        try {
            val query = "INSERT INTO clientes (nombre, apellidos, creado_el, actualizado_el) VALUES (?, ?, NOW(), NOW())"
            val statement = connection.prepareStatement(query, java.sql.Statement.RETURN_GENERATED_KEYS)
            statement.setString(1, nombre)
            statement.setString(2, apellidos)
            
            val affectedRows = statement.executeUpdate()
            
            if (affectedRows > 0) {
                val generatedKeys = statement.generatedKeys
                if (generatedKeys.next()) {
                    newId = generatedKeys.getInt(1)
                }
                generatedKeys.close()
            }
            statement.close()
            connection.close()
        } catch (e: SQLException) {
            android.util.Log.e("ClienteDAOImpl", "SQL Error", e)
            e.printStackTrace()
        }
        return newId
    }
}